ethereum/p2p/src/test/java/tech/pegasys/pantheon/ethereum/p2p/discovery/internal/RecursivePeerRefreshStateTest.java

```diff
@@ -77,50 +96,348 @@
+    peer_000.setStatus(PeerDiscoveryStatus.BONDED);
+
+    recursivePeerRefreshState.onBondingComplete(peer_000);
+
+    verify(neighborFinder).findNeighbours(peer_000, TARGET);
+
+    recursivePeerRefreshState.onNeighboursPacketReceived(peer_000, neighborsPacketData_000);
+
+    verify(bondingAgent).performBonding(peer_010);
+    verify(bondingAgent).performBonding(peer_011);
+    verify(bondingAgent).performBonding(peer_012);
+    verify(bondingAgent).performBonding(peer_013);
+
+    peer_010.setStatus(PeerDiscoveryStatus.BONDED);
+    peer_011.setStatus(PeerDiscoveryStatus.BONDED);
+    peer_012.setStatus(PeerDiscoveryStatus.BONDED);
+    peer_013.setStatus(PeerDiscoveryStatus.BONDED);
+
+    recursivePeerRefreshState.onBondingComplete(peer_010); // Trigger the next round...
+
+    verify(neighborFinder, never()).findNeighbours(peer_010, TARGET);
+    verify(neighborFinder).findNeighbours(peer_011, TARGET);
+    verify(neighborFinder).findNeighbours(peer_012, TARGET);
+    verify(neighborFinder).findNeighbours(peer_013, TARGET);
+
+    timerUtil.runTimerHandlers();
+
+    verify(neighborFinder).findNeighbours(peer_010, TARGET);
+
+    timerUtil.runTimerHandlers();
+
+    recursivePeerRefreshState.onNeighboursPacketReceived(peer_011, neighborsPacketData_011);
+
+    verify(bondingAgent).performBonding(peer_120);
+    verify(bondingAgent).performBonding(peer_121);
+    verify(bondingAgent).performBonding(peer_122);
+    verify(bondingAgent).performBonding(peer_123);
+  }
+
+  @Test
+  public void shouldStopWhenAllNodesHaveBeenQueried() {
+    setupRelativePeerTree();
+
+    recursivePeerRefreshState.start(singletonList(peer_000), TARGET);
+
+    verify(bondingAgent).performBonding(peer_000);
+
+    peer_000.setStatus(PeerDiscoveryStatus.BONDED);
+
+    recursivePeerRefreshState.onBondingComplete(peer_000);
+
+    verify(neighborFinder).findNeighbours(peer_000, TARGET);
+
+    recursivePeerRefreshState.onNeighboursPacketReceived(peer_000, neighborsPacketData_000);
+
+    verify(bondingAgent).performBonding(peer_010);
+    verify(bondingAgent).performBonding(peer_011);
+    verify(bondingAgent).performBonding(peer_012);
+    verify(bondingAgent).performBonding(peer_013);
+
+    peer_010.setStatus(PeerDiscoveryStatus.BONDED);
+    peer_011.setStatus(PeerDiscoveryStatus.BONDED);
+    peer_012.setStatus(PeerDiscoveryStatus.BONDED);
+    peer_013.setStatus(PeerDiscoveryStatus.BONDED);
+
+    recursivePeerRefreshState.onBondingComplete(peer_010); // Trigger the next round...
+
+    verify(neighborFinder, never()).findNeighbours(peer_010, TARGET);
+    verify(neighborFinder).findNeighbours(peer_011, TARGET);
+    verify(neighborFinder).findNeighbours(peer_012, TARGET);
+    verify(neighborFinder).findNeighbours(peer_013, TARGET);
+
+    timerUtil.runTimerHandlers();
+
+    verify(neighborFinder).findNeighbours(peer_010, TARGET);
+
+    verifyNoMoreInteractions(bondingAgent);
+  }
+
+  @Test
+  public void shouldStopWhenMaximumNumberOfRoundsReached() {}
```

Comment:
This one, when implemented, will need more peers but probably best to just create them on the fly like we did with peer1, 2 and 3 (probably extract a method to create the whole `DiscoveryPeer` instance given an ID.